### PR TITLE
split _reduce to _xReduce(for transformers) & _reduce(for reducers)

### PR DESCRIPTION
**Pull request by xgbuils**
Following the approach about what I said in #3247, I also extracted _arrayReduce and _xArrayReduce to be used in places where function are only working with array lists. In this way, we are not executing additional type checks in cases are not needed.

### CODE REVIEW

**Comment by xgbuils** [comment]
Notice that this PR doesn't colide with #2767. #2767 is a behavior change. This PR it's just a refactor.

If we merge this PR and we want to apply the behavior of #2767 we just need to remove the lines:

``` javascript
if (typeof list['fantasy-land/reduce'] === 'function') {
  return list['fantasy-land/reduce'](fn, acc);
}
```
in `_createReducer.js`

and adding in `_reduce.js`:
``` javascript
var _baseReduce = _createReduce(_arrayReduce, _methodReduce, _iterableReduce);
export default function _reduce(reducer, acc, list) {
  if (typeof list['fantasy-land/reduce'] === 'function') {
    return list['fantasy-land/reduce'](fn, acc);
  }
  return _baseReduce(reducer, acc, list);
}
```

**Comment by xgbuils** [comment]
Hi! Some benchmarks changin map.bench.js

``` javascript
var obj = {a: 1, b: 2};
...
    'map(sq, obj)': function() {
      map(sq, obj);
    },
```
Original:

**Comment by xgbuils** [comment]
Hi! Some benchmarks adding benchs for objects in map.bench.js

``` javascript
var obj = {a: 1, b: 2};
...
    'map(sq, obj)': function() {
      map(sq, obj);
    },
```
Original:
```
┌────────────────────────┬────────────────────────┬────────────────────────┐
│ map                    │ Hertz                  │ Margin of Error        │
├────────────────────────┼────────────────────────┼────────────────────────┤
│ map(sq, obj)           │ 956,837                │ 4.81%                  │
└────────────────────────┴────────────────────────┴────────────────────────┘
```
After change:
```
┌────────────────────────┬────────────────────────┬────────────────────────┐
│ map                    │ Hertz                  │ Margin of Error        │
├────────────────────────┼────────────────────────┼────────────────────────┤
│ map(sq, obj)           │ 2,099,464              │ 6.11%                  │
└────────────────────────┴────────────────────────┴────────────────────────┘
```

Using more keys implies less difference in gain of performance:
```
var nums = [8, 2, 85, 2, 34, 3, 23, 247, 57, 8, 0, 6, 5, 46, 54, 643];
var obj = {...nums};
...
    'map(sq, obj)': function() {
      map(sq, obj);
    },
```
Original:
```
┌────────────────────────┬────────────────────────┬────────────────────────┐
│ map                    │ Hertz                  │ Margin of Error        │
├────────────────────────┼────────────────────────┼────────────────────────┤
│ map(sq, obj)           │ 571,880                │ 0.41%                  │
└────────────────────────┴────────────────────────┴────────────────────────┘
```
After change:
```
┌────────────────────────┬────────────────────────┬────────────────────────┐
│ map                    │ Hertz                  │ Margin of Error        │
├────────────────────────┼────────────────────────┼────────────────────────┤
│ map(sq, obj)           │ 778,917                │ 0.42%                  │
└────────────────────────┴────────────────────────┴────────────────────────┘
```

File: source/ap.js [hunk 1]
```diff
@@ -38,7 +38,11 @@ var ap = _curry2(function ap(applyF, applyX) {
         ? applyF.ap(applyX)
         : typeof applyF === 'function'
           ? function(x) { return applyF(x)(applyX(x)); }
-          : _reduce(function(acc, f) { return _concat(acc, map(f, applyX)); }, [], applyF)
+          : _reduce(
+            function(acc, f) { return _concat(acc, map(f, applyX)); },
+            [],
+            applyF
+          )
```
Comment (CrossEye) [inline source/ap.js:45]: Post `1.0`, I would like to do some auto-formatting, using Prettier or something, and from then on simply ignore such things.  The trouble, of course, is that auto-formatters are mostly designed around OOP/imperative styles, and those don't always transfer well to FP/declarative code.  Still, I would love to simply never have to make such judgement calls.

**Review by CrossEye — APPROVED** [review]
This looks great to me.  In the end the code is cleaner, *and* it's more performant.  It's hard to beat that!

:herb:

File: source/ap.js (hunk 1 above)
Comment (xgbuils) [reply source/ap.js:45]: My bad. I'm gonna revert this formatting part.

**Comment by xgbuils** [comment]
internal `_reduce` implementation changed and now only works with standard reducers. The new internal `_xReduce` function only works with transformers. Then, some `_reduce` usages are hidden and difficult to review because the files didn't change but the implempementation of `_reduce` did. I'm gonna list the usages of _reduce:

- [ap.js](https://github.com/xgbuils/ramda/blob/refactor-3247/source/ap.js#L41)
- [collectBy.js](https://github.com/xgbuils/ramda/blob/refactor-3247/source/collectBy.js#L36)
- [count.js](https://github.com/xgbuils/ramda/blob/refactor-3247/source/count.js)
- [pipeWith.js](https://github.com/xgbuils/ramda/blob/refactor-3247/source/pipeWith.js#L42)

**Review by CrossEye — APPROVED** [review]

**Review by customcommander — APPROVED** [review]

**Pull request merged**